Implement renaming dependencies in the manifest

alexcrichton opened the pull request:
This commit implements a new unstable feature for manifests which allows
renaming a crate when depending on it. For example you can do:

```toml
cargo-features = ["dependencies-as"]

...

[dependencies]
foo = "0.1"
bar = { version = "0.1", registry = "custom", package = "foo" }
baz = { git = "https://github.com/foo/bar", package = "foo" }
```

Here three crates will be imported but they'll be made available to the Rust
source code via the names `foo` (crates.io), `bar` (the custom registry), and
`baz` (the git dependency). The *package* name, however, will be `foo` for all
of them. In other words the git repository here would be searched for a crate
called `foo`. For example:

```rust
extern crate foo; // crates.io
extern crate bar; // registry `custom`
extern crate baz; // git repository
```

The intention here is to enable a few use cases:

* Enable depending on the same named crate from different registries
* Allow depending on multiple versions of a crate from one registry
* Removing the need for `extern crate foo as bar` syntactically in Rust source

Currently I don't think we're ready to stabilize this so it's just a nightly
feature, but I'm hoping we can continue to iterate on it!

cc #1311

rust-highfive commented:
r? @matklad

(rust_highfive has picked a reviewer for you, use r? to override)

sfackler commented:
This may have already been discussed elsewhere, but I think it would be a bit less confusing to do something like this:

```toml
[dependencies]
foo = "0.1"
bar = { version = "0.1", registry = "custom", name = "foo" }
baz = { git = "https://github.com/foo/bar", name = "foo" }
```

It avoids the extra array, and makes it so that the set of names you're importing (modulo crate/lib naming differences) are all in the same place.

alexcrichton commented:
Heh that's actually the first syntax I thought of as well, although it was changed due to https://github.com/rust-lang/rfcs/pull/2126#issuecomment-328921126, namely:

> the desired crate alias as the leading key in Cargo.toml conflates packages and crates (a common conflation that we should probably work on separately).

For example today you actually specify *package* names as the key in `[dependencies]` whereas the `as`/alias business we want is to rename the crate. (it's possible for packages to be named differently from their library crate).

That being said I also just realized something that can be troubling. How, for example, do we enable features for these dependencies? For example:

```toml
[dependencies]
foo = [
   { version = "0.1" },
   { version = "0.1", registry = "custom", as = "bar" },
   { git = 'https://github.com/foo/bar, as = "baz" },
]

[features]
some-feature = ["foo/another-feature"] # what version of `foo`?
```

This I think is much clearer with what @sfackler is thinking:

```toml
[dependencies]
foo = "0.1"
bar = { version = "0.1", registry = "custom", name = "foo" }
baz = { git = "https://github.com/foo/bar", name = "foo" }

[features]
some-feature = ["foo/another-feature"] # aha, the crates.io version!
```

That may tip my own opinion towards the latter!

sfackler commented:
Oh yeah, good point. It's too bad you're allowed to set the library name separately from the crate name :(

matklad commented on src/cargo/core/features.rs:
```diff
@@ -125,6 +125,9 @@ features! {
         // Downloading packages from alternative registry indexes.
         [unstable] alternative_registries: bool,
+
+        // Downloading packages from alternative registry indexes.
```
Wrong comment

matklad commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -931,28 +931,44 @@ Cargo.toml. This warning might turn into a hard error in the future.",
+                .filter_map(|d| d.rename())
+                .next();
+
+            v.push(name.unwrap_or(&dep.target.crate_name()));
```
Could we make this `.unwrap_or` an utility method of something? Like `fn crate_name(&self) -> String`, which deals with `as` and, probably, with translating `-` to `_` as well.

matklad commented on tests/rename-deps.rs:
```diff
@@ -0,0 +1,174 @@
+            bar = [
+                { version = "0.1.0" },
+                { version = "0.2.0", as = "baz" },
+            ]
```
Hm the syntax does not look too natural to me. Is the following a valid TOML?

```TOML
[dependencies]
bar = "0.1.0"
bar = { version = "0.2.0", as = "baz"}
```

alexcrichton replied:
Unfortunately TOML-the-spec doesn't allow for something like that b/c it's counted as a redefinition of an existing key :(

bors commented:
:umbrella: The latest upstream changes (presumably #4844) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #4957) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #5011) made this pull request unmergeable. Please resolve the merge conflicts.

retep998 commented:
If we wanted to make it clear where we're specifying the package name vs the crate name, perhaps we could do something like...
```toml
[dependencies]
foo = "0.1"
bar = { version = "0.1", registry = "custom", package = "foo" }
baz = { git = "https://github.com/foo/bar", package = "foo" }

[features]
some-feature = ["foo/another-feature"] # aha, the crates.io version!
```

alexreg commented:
@alexcrichton Great work on this. I think this feature is very much due. Funnily enough, I was just going to suggest the key `package` (or `pkg`), but @retep998 just beat me to it. The semantics would be changed from your original approach, and the current method, with keys now representing user-facing crate names rather than package names, but it would at least be operationally backwards-compatible (in the case of no `package` value, that is). In other words, the value of the `package` field would default to the user-facing crate name (i.e. alias). This seems quite nice to me.

It's also worth noting that this PR helps with https://github.com/rust-lang/rust/issues/44660.

Finally, maybe add a "fixes https://github.com/rust-lang/cargo/issues/1311" to your original PR message?

alexcrichton commented:
I've updated with the feedback so far (and switched to the non-array syntax using a `package` key), re-r? @matklad 


@alexreg oh we typically like to keep the issues open to track the feature, but I should have cc'd it in the original message for sure!

alexreg commented:
@alexcrichton Yeah, that's fair enough. Anyway, the new semantics are with the alias on the LHS of the `=`, right? (i.e. what I elaborated on above.)

alexcrichton commented:
@alexreg, correct!

alexreg commented:
Super. Hopefully @matklad can review soon and then we can get this merged.

alexcrichton commented:
@matklad sure yeah, although for specifying one crate twice we're already pretty bad about that today unfortunately. For example something like:


```toml
[dependencies]
foo = "0.1"

[target.'cfg(unix)'.dependencies]
foo = "0.1"
```

is accepted and works, so I figure we can probably just let rustc sort it out if it's specified multiple times.

alexcrichton commented on src/cargo/core/dependency.rs:
```diff
@@ -27,6 +27,7 @@ struct Inner {
     specified_req: bool,
     kind: Kind,
     only_match_name: bool,
+    rename: Option<String>,
```
Oh this is slightly different through b/c it's not actually the package, but rather what the dependency is renaming the crate to

alexreg replied:
Yeah, even though there's a one-to-one correspondence with the `package` attribute, `rename` probably makes more sense in code (perhaps with a comment?)... that's my thought, anyway.

matklad commented:
@bors r+

bors commented:
:pushpin: Commit 79942fe has been approved by `matklad`

<!-- @bors r=matklad 79942fea1bb7a480f6eebd6cd1f8a04ac9f45a0b -->

alexreg commented:
Good work with this @acrichto! Thanks. 👍🏻

alexreg commented:
Good work with this @alexcrichton! Thanks. 👍🏻

bors commented:
:hourglass: Testing commit 79942fea1bb7a480f6eebd6cd1f8a04ac9f45a0b with merge 4f1c6dba4878a578a21741582741998593d5605b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.3643), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/343538107?utm_source=github_status&utm_medium=notification)
Approved by: matklad
Pushing 4f1c6dba4878a578a21741582741998593d5605b to master...

peterjoel commented:
@matklad wrote:
>Using exactly the same library twice. This presumably should be an error


Probably should be a warning - and maybe not even that. Generated code shouldn't have to worry about this edge case. But two references to the same exact create should only result in the crate being included once, just with two aliases.

aaronsky commented:
With this merged, does this mean that the item under #44931 should be updated? Apologies if this is the wrong place to ask about this

alexreg commented:
@aaronsky Yep, it should be. Maybe ping @aturon on that thread to update the checkbox(es). :-)

aldanor commented:
Here's a curious (but not so improbable) edge case with proc macros. Wondering how should this be handled? (or am I misunderstanding something)

- There's three crates, `foo`, `foo-derive` and `bar`
- `foo-derive` is a proc-macro crate which depends on crate `foo` (without any renaming) and mentions `::foo::Foo` in the resulting token stream
- `bar` is a top-level crate that depends on both `foo` and `foo-derive`
- Using the rename feature, in `bar`'s manifest we have `baz = { package = "foo" }`

Now the derive proc macro seems to be broken upon use, like so:

```rust
^^^^^^ Could not find `foo` in `{{root}}`
```

What would be a proper way to refer to it then?..

aldanor commented:
Here's a concrete example, using [`num-derive`](https://github.com/rust-num/num-derive) as an example.

`Cargo.toml`:

```toml
cargo-features = ["rename-dependency"]

[package]
name = "rename-test"
version = "0.1.0"
edition = "2018"

[dependencies]
foo = { package = "num-traits", version = "0.2" }
num-derive = "0.2"
```

`src/lib.rs`:

```rust
use num_derive::{FromPrimitive, ToPrimitive};

#[derive(FromPrimitive, ToPrimitive)]
enum Color { Red, Blue, Green }
```

... which results in...

```
error[E0463]: can't find crate for `_num_traits`
 --> src/lib.rs:3:25
  |
3 | #[derive(FromPrimitive, ToPrimitive)]
  |                         ^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
```

alexcrichton commented:
@aldanor that's a general problem with hygiene and procedural macros, and most procedural macros only work if the corresponding library crate is imported with one precise name (and not renamed). There's no way, for example, to get importing the `serde` crate under a different name to work with its derive modes.

ehuss commented:
> maybe there should be some crate-resolving syntax

There is an issue open for this at rust-lang/rust#54363.

updogliu commented:
Can this enable depending on the same version of a crate but using conflicting features at the same time (with different renaming of course)?

ehuss commented:
@updogliu A package cannot have two dependencies to the same version of a crate.

updogliu commented:
@ehuss What I am trying to achieve is using a crate of the same version but with conflicting features in two members of a workspace. Is that a valid motivation?

ehuss commented:
If the two members are built separately, then they will each use a different copy of the dependent crate (with different features activated).  If the two members are being linked together, then it will unify the features and use a single copy.  Conflicting or mutually exclusive features aren't really supported.  There are a number of issues in the tracker here to make this more flexible, but we're still in the design phase.

updogliu commented:
@ehuss I have an example at [here](https://github.com/updogliu/play/tree/master/ws_features). There are three members in a workspace: `foo`, `bar`, `common`. `foo` and `bar` do not depend on each other, but they both depend on `common` and using mutually exclusive features. Is this kind of use to be supported or not?

retep998 commented:
Mutually exclusive features are not supported.

updogliu commented:
@retep998 I understand that it is currently not supported. But since this nice renaming dependencies feature is added to support using different versions of a crate, maybe it can be augmented a bit to support exclusive features under different renaming as well?

updogliu commented:
Oh, there is already an example in the doc
https://doc.rust-lang.org/cargo/reference/manifest.html#the-features-section

burdges commented:
I presume cargo features only work on nightly installs
